### PR TITLE
[Remake] War Ops (TC boost for war declaration)

### LINKED ISSUE #9560: Need proper war-ops with extra tc and annoucements
## Description

**Reproduction**

**Screenshots**

**Additional context**

### PR DESCRIPTION
**Pull request by Morb0**
## About the PR

Original PR #14016

Gives opportunity to get more TC for war declaration with withdraw that they must wait 15 minutes after it (so crew can prepare)

These conditions must be met for war declaration

- More than 3 nuclear operatives
- Nothing traveled by FTL from syndie base by FTL
- Less than 6 minutes passed from round start
- War Ops are enabled on the server

After war declaration the FTL from syndie outpost will be disabled for 15 minutes. This gives crew time to prepare. Crew also can't call an evacuation shuttle and leave from the station in that delay.

Syndie comms console are deleted from outpost.

There is new item called "the declaration of war". It spawns in commander hand. You can declare war with it and customize its message (it have default one) and uses `TextEdit` as other announcements now. It also shows current status about war:
- Is conditions are met (and which one if not)
- How much time left for declaration
- How much time left for FTL enable (if war was declared and you are waiting for it)

The boost amount is +40 TC for each operative. You can get +200 TC in total for declaration. So full war ops crew have 400 TC for their stuff.

## Why / Balance

Closes #9560

## Technical details

- `WarConditionOnExamine` was not added to this PR, it seemed redundant.
- Adds `AddedAt` (TimeSpan) field to `GameRuleComponent` because it can be useful for new game rules to calculate rule active time.
- Adds not best methods `TryGetRuleFromOperative` and `TryGetRuleFromGrid`, I really don't know how best to handle getting a game rule when we can have many of them at once.
- Adds `ShuttleConsoleFTLTravelStartEvent` event that emits console after FTP was approved, existing `FTLStartedEvent` emits with delay, so we can't block ops from starting FTL and not to claim war with existed event.

## Media

https://github.com/space-wizards/space-station-14/assets/14136326/ce216b48-0250-4859-99ef-c756818e3b4a

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

**Changelog**

:cl: CaptainSqrBeard, Morb0
- add: Nuclear operative commanders may now elect to declare war, earning their team a TC boost.

### CODE REVIEW

**Comment by JoeHammad1844** [comment]
nice seth reference

**Comment by JoeHammad1844** [comment]
also the scanning animation on the war declarer doesn't seem to go back and forth, like I assume the intention is, it just teleports to the right

**Comment by Flareguy** [comment]
It is not meant to go back and forth (I am the one who made the sprite.) It DOES animate too fast, however - here's what the original gif export of the sprite looks like:

it's boob

**Comment by JoeHammad1844** [comment]
nice 👍

**Comment by Just-a-Unity-Dev** [comment]
should be a bit more than 20 seconds for nukies to discuss "oh do we stealthies or warops"

**Comment by nopeingeneer** [comment]
> should be a bit more than 20 seconds for nukies to discuss "oh do we stealthies or warops"

I mean it says that nukies got 6 min before war device run out of time. It enough i think

**Comment by nopeingeneer** [comment]
This is probably dead™️

File: Content.Server/GameTicking/Rules/Components/GameRuleComponent.cs [hunk 1]
```diff
@@ -14,6 +14,12 @@ public sealed class GameRuleComponent : Component
     [DataField("active")]
     public bool Active;
 
+    /// <summary>
+    /// Game time when game rule was activated
+    /// </summary>
+    [DataField("activatedAt")]
+    public TimeSpan ActivatedAt;
```
Comment (metalgearsloth) [inline Content.Server/GameTicking/Rules/Components/GameRuleComponent.cs:21]: TimeOffsetSerializer.

File: Content.Server/GameTicking/Rules/Components/NukeopsRuleComponent.cs [hunk 2]
```diff
@@ -44,6 +41,54 @@ public sealed class NukeopsRuleComponent : Component
...
+    /// Whether or not nukie left their outpost
+    /// </summary>
+    [DataField("leftOutpost")]
+    public bool LeftOutpost = false;
+
+    /// <summary>
+    ///     Enables opportunity to get extra TC for war declaration
+    /// </summary>
+    [DataField("enableWarOps")]
+    public bool EnableWarOps = true;
```
Comment (metalgearsloth) [inline Content.Server/GameTicking/Rules/Components/NukeopsRuleComponent.cs:54]: Call it CanEnableWarOps/

File: Content.Server/GameTicking/Rules/Components/NukeopsRuleComponent.cs [hunk 3]
```diff
@@ -44,6 +41,54 @@ public sealed class NukeopsRuleComponent : Component
...
+
+    /// <summary>
+    ///     Indicates time when war has been declared, null if not declared
+    /// </summary>
+    [DataField("warDeclaredTime")]
+    public TimeSpan? WarDeclaredTime;
```
Comment (metalgearsloth) [inline Content.Server/GameTicking/Rules/Components/NukeopsRuleComponent.cs:60]: TimeOffsetSerializer.

File: Content.Server/GameTicking/Rules/Components/NukeopsRuleComponent.cs [hunk 4]
```diff
@@ -44,6 +41,54 @@ public sealed class NukeopsRuleComponent : Component
...
+    /// </summary>
+    [DataField("warNukieArriveDelay")]
+    public TimeSpan WarNukieArriveDelay = TimeSpan.FromMinutes(15);
+
+    /// <summary>
+    ///     Minimal operatives count for war declaration
+    /// </summary>
+    [DataField("warDeclarationMinOpsSize")]
+    public int WarDeclarationMinOps = 4;
+
```
Comment (metalgearsloth) [inline Content.Server/GameTicking/Rules/Components/NukeopsRuleComponent.cs:91]: Datafields should match field name unless good reason not to.

**Review by EmoGarbage404 — APPROVED** [review]

File: Content.Server/GameTicking/Rules/Components/NukeopsRuleComponent.cs [hunk 5]
```diff
@@ -44,6 +41,54 @@ public sealed class NukeopsRuleComponent : Component
...
+    ///     Indicates time when war has been declared, null if not declared
+    /// </summary>
+    [DataField("warDeclaredTime")]
+    public TimeSpan? WarDeclaredTime;
+
+    /// <summary>
+    ///     Delay between declaration of war and enabling FTL at outpost map
+    /// </summary>
+    [DataField("warShuttleDelay")]
+    public bool PreventShuttleInDelay = true;
```
Comment (metalgearsloth) [inline Content.Server/GameTicking/Rules/Components/NukeopsRuleComponent.cs:66]: Can't you just make WarNukieArriveDelay a nullable timespan instead and remove this?

File: Content.Server/GameTicking/Rules/NukeopsRuleSystem.cs [hunk 6]
```diff
@@ -854,6 +990,78 @@ private void OnStartAttempt(RoundStartAttemptEvent ev)
...
+                var timeAfterDeclaration = _gameTiming.CurTime.Subtract(nukeops.WarDeclaredTime.Value);
+                var timeRemain = nukeops.WarNukieArriveDelay.Subtract(timeAfterDeclaration);
+                if (timeRemain > TimeSpan.Zero)
+                {
+                    ev.Cancelled = true;
+                    ev.Reason = Loc.GetString("war-ops-infiltrator-unavailable", ("minutes", timeRemain.Minutes), ("seconds", timeRemain.Seconds));
+                }
+            }
+        }
+    }
```
Comment (metalgearsloth) [inline Content.Server/GameTicking/Rules/NukeopsRuleSystem.cs:1018]: Isn't this raised on anything FTLing?

File: Content.Server/GameTicking/Rules/NukeopsRuleSystem.cs [hunk 7]
```diff
@@ -854,6 +990,78 @@ private void OnStartAttempt(RoundStartAttemptEvent ev)
...
+            if (mapOutpost == mapShuttle)
+            {
+                nukeops.LeftOutpost = true;
+
+                var gridUid = Transform(ev.Uid).GridUid;
+                if (gridUid != null && TryGetRuleFromGrid(gridUid.Value, out var comps))
+                    _warDeclaratorSystem.RefreshAllUI(comps.Value.Item1, comps.Value.Item2);
+            }
+        }
+    }
```
Comment (metalgearsloth) [inline Content.Server/GameTicking/Rules/NukeopsRuleSystem.cs:1043]: Same here.

File: Content.Server/NukeOps/WarDeclaratorComponent.cs [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    /// <summary>
+    /// Permission to customize message text
+    /// </summary>
+    [ViewVariables(VVAccess.ReadWrite)]
+    [DataField("allowEditMessage")]
+    public bool AllowEditMessage { get; set; } = true;
+
+    [ViewVariables(VVAccess.ReadWrite)]
+    [DataField("maxMessageLength")]
+    public int MaxMessageLength { get; set; } = 512;
```
Comment (metalgearsloth) [inline Content.Server/NukeOps/WarDeclaratorComponent.cs:27]: Use fields not properties.

File: Content.Server/NukeOps/WarDeclaratorComponent.cs [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    [ViewVariables(VVAccess.ReadWrite)]
+    [DataField("message")]
+    public string Message { get; set; } = "war-declarator-default-message";
+
+    /// <summary>
+    /// Permission to customize message text
+    /// </summary>
+    [ViewVariables(VVAccess.ReadWrite)]
+    [DataField("allowEditMessage")]
+    public bool AllowEditMessage { get; set; } = true;
```
Comment (metalgearsloth) [inline Content.Server/NukeOps/WarDeclaratorComponent.cs:23]:
```suggestion
    /// <summary>
    /// Permission to customize message text
    /// </summary>
    [ViewVariables(VVAccess.ReadWrite)]
    [DataField("allowEditingMessage")]
    public bool AllowEditingMessage = true;
```

File: Content.Server/NukeOps/WarDeclaratorSystem.cs [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+using Content.Shared.Database;
+using Content.Shared.Interaction;
+using Content.Shared.NukeOps;
+using Robust.Server.GameObjects;
+
+namespace Content.Server.NukeOps;
+
+public sealed class WarDeclaratorSystem : EntitySystem
+{
+    [Dependency] private readonly UserInterfaceSystem _userInterfaceSystem = default!;
```
Comment (metalgearsloth) [inline Content.Server/NukeOps/WarDeclaratorSystem.cs:14]: Needs documenting.

File: Content.Server/NukeOps/WarDeclaratorSystem.cs [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+        if (!_nukeopsRuleSystem.TryGetRuleFromOperative(args.User, out var comps))
+        {
+            var msg = Loc.GetString("war-declarator-not-nukeops");
+            _popupSystem.PopupEntity(msg, uid);
+            return;
+        }
+
+        _userInterfaceSystem.TryOpen(uid, WarDeclaratorUiKey.Key, actor.PlayerSession);
+        UpdateUI(uid, comps.Value.Item1, comps.Value.Item2);
+    }
```
Comment (metalgearsloth) [inline Content.Server/NukeOps/WarDeclaratorSystem.cs:42]: Isn't this just an ActivatableUI?

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Server/GameTicking/Rules/NukeopsRuleSystem.cs (hunk 6 above)
Comment (Morb0) [reply Content.Server/GameTicking/Rules/NukeopsRuleSystem.cs]: Add check for grid

File: Content.Server/NukeOps/WarDeclaratorSystem.cs (hunk 11 above)
Comment (Morb0) [reply Content.Server/NukeOps/WarDeclaratorSystem.cs]: Use ActivatableUI, but popup that owner not nukeops lost

Comment (Morb0) [reply Content.Server/NukeOps/WarDeclaratorSystem.cs]: Okay, returned ops check and UI update using `ActivatableUIOpenAttemptEvent`

**Comment by CaptainSqrBeard** [comment]
I was not expected that someone would remake my code. Thank you @Morb0

**Comment by Partmedia** [comment]
Thank you for your contribution. I have reviewed the code, so this makes 3 code reviews in all. I've done some testing in-game and found that:

- Late join nukies do not receive their TC bonus after declaring war
- Late join nukies do not change the ability to declare war, unless the commander closes their war declarator UI and re-opens it
- ~~Adding a shuttle console on the main station allows the station to call the emergency shuttle~~ (submitter fixed this before I finished this post)
- Declaring war with an empty message does not work (i.e. send the default message)

None of these are significant issues that would block a merge.

**Review by Partmedia — APPROVED** [review]

**Comment by JoeHammad1844** [comment]
HALAL!!

**Comment by metalgearsloth** [comment]
> HALAL!!

Please stop posting inane comments everywhere.

**Comment by JoeHammad1844** [comment]
expressing excitement is now illegal

**Comment by metalgearsloth** [comment]
I don't want to scroll through spam on github channel to get to prs I might need to review.

**Pull request merged**